File: java/A1150-Travelling Salesman Problem.java
//只对了2号测试点。。。（一共4个）
import java.util.Scanner;

public class Main {

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		int m = scanner.nextInt();
		int[][] map = new int[n+1][n+1];
		for(int i=0;i<m;i++) {
			int c1 = scanner.nextInt();
			int c2 = scanner.nextInt();
			int d = scanner.nextInt();
			map[c1][c2]=map[c2][c1]=d;
		}
		int k = scanner.nextInt();
		int shortestDist = Integer.MAX_VALUE;
		int shortestPath = -1;
		loop:
		for(int i=0;i<k;i++) {
			int x = scanner.nextInt();
			int start = scanner.nextInt();
			int backup = start;
			int end = -1;
			int totalDist = 0;
			boolean[] visit = new boolean[n+1];
			int notSimple = 0;
			System.out.print("Path "+(i+1)+": ");
			for(int j=1;j<x;j++) {
				end = scanner.nextInt();
				if(visit[end]) notSimple++;
				else visit[end] = true;
				
				if(map[start][end]==0) {
					System.out.println("NA (Not a TS cycle)");
					totalDist = Integer.MAX_VALUE;
					continue loop;
				}
				else {
					totalDist += map[start][end];
					if(j==x-1&&backup!=end) {
						System.out.println(totalDist+" (Not a TS cycle)");
						continue loop;
					}
				}
				
				start = end;
			}
			
			if(notSimple==0 && x==n+1) {
				System.out.println(totalDist+" (TS simple cycle)");
				if(totalDist<shortestDist) {
					shortestDist = totalDist;
					shortestPath = i+1;
				}
			}
			else if(x-notSimple==n+1){
				System.out.println(totalDist+" (TS cycle)");
				if(totalDist<shortestDist) {
					shortestDist = totalDist;
					shortestPath = i+1;
				}
			}
			else {
				System.out.println(totalDist+" (Not a TS cycle)");
			}
		}
		System.out.println("Shortest Dist("+shortestPath+") = "+shortestDist);
		
		scanner.close();
	}
}

File: C++/A1150-Travelling Salesman Problem.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>
#include <algorithm>

//前两个测试点过不去
int main() {
	int n;
	int m;
	scanf_s("%d",&n);
	scanf_s("%d",&m);
	int map[201][201] = {0};
	int c1;
	int c2;
	int d;
	for (int i = 0; i < m; i++) {
		scanf_s("%d", &c1);
		scanf_s("%d", &c2);
		scanf_s("%d", &d);
		map[c1][c2] = map[c2][c1] = d;
	}
	int k;
	scanf_s("%d", &k);
	int shortestDist = 0x3f3f3f3f;
	int shortestPath = -1;

	for (int i = 0; i < k; i++) {
		int x;
		scanf_s("%d", &x);
		int start;
		scanf_s("%d", &start);
		int backup = start;
		int end = -1;
		int totalDist = 0;
		bool visit[201] = { false };
		int notSimple = 0;
		std::cout<<"Path "<<(i+1)<<": ";
		bool flag = false;
		for (int j = 1; j < x; j++) {
			scanf_s("%d", &end);
			if (visit[end]) notSimple++;
			else visit[end] = true;

			if (map[start][end] == 0) {
				flag = true;
				totalDist = 0x3f3f3f3f;
				break;
			}
			totalDist += map[start][end];
			start = end;
		}
		if (flag) {
			std::cout << "NA (Not a TS cycle)"<<std::endl;
		}
		else if(backup != end){
			std::cout <<totalDist<< " (Not a TS cycle)"<<std::endl;
		}
		else if (notSimple == 0 && x == n + 1) {
			std::cout << totalDist << " (TS simple cycle)" <<std::endl;
			if (totalDist < shortestDist) {
				shortestDist = totalDist;
				shortestPath = i + 1;
			}
		}
		else if (x - notSimple == n + 1) {
			std::cout << totalDist << " (TS cycle)" <<std::endl;
			if (totalDist < shortestDist) {
				shortestDist = totalDist;
				shortestPath = i + 1;
			}
		}
		else {
			std::cout << totalDist << " (Not a TS cycle)"<<std::endl;
		}
	}
	std::cout << "Shortest Dist("<< shortestPath <<") = "<<shortestDist<<std::endl;

	getchar();
	getchar();

	return 0;
}